packages/vm/src/context.rs

```diff
@@ -205,68 +245,97 @@ mod iter_support {
 
 /** context data **/
 
-struct ContextData<S: Storage> {
+struct ContextData<S: Storage, Q: Querier> {
     data: Option<S>,
+    querier: Option<Q>,
```

Comment:
Is this really needed here? It pollutes all the storage handling. Isn't Querier more like Api than like Storage?

Reply:
This is the big question.

To act like api, I need to implement Copy and Clone.
There is state there behind it.

However,  &Querier (non mutable) is Copy and Cloneable.

I could try a bigger refactor there, but this would probably touch all usages of Extern and I would rather try that after this PR.

Reply:
Look at the constructor types:

```rust
impl<S, A, Q> Instance<S, A, Q>
where
    S: Storage + 'static,
    A: Api + 'static,
    Q: Querier + 'static,
{
    pub fn from_code(code: &[u8], deps: Extern<S, A, Q>, gas_limit: u64) -> Result<Self> {
```

I don't see a way to get `&Q` to implement `'static`

Reply:
So this state behind `Querier` comes from the mock implementation, right?

```rust
/// A stateless convenience wrapper around imports provided by the VM
#[derive(Copy, Clone)]
pub struct ExternalQuerier {}
```

```rust
/// MockQuerier holds an immutable table of bank balances
/// TODO: also allow querying contracts
#[derive(Clone)]
pub struct MockQuerier {
    balances: HashMap<HumanAddr, Vec<Coin>>,
}
```

I know that it is a mayor annoyance to try using something without `Copy` in the import implementations. I tried to remove Copy from the Api trait once and failed horribly.

I see two ways here:
1. Find a way to implement the data in `MockQuerier` in a `Copy`able fashion
2. Find a way to clone api and querier into import implementations

I prefer 2., as it keeps us more flexible. But using the same overhead for Querier as the non-Clonable Storage seems not very elegant.

Reply:
I remember the problem with cloning into import implementations: The callbacks need to be `Fn`, not `FnOnce`, i.e. they can be called multiple times. When you move a previously cloned value into the callback, it only exists for one function execution. And you cannot use a reference, since there is the function lives longer than the objects of the that are passed in.

Api only works because it can be copied without any lifetime restrictions.

The only place for long living data is `Ctx.data`, so the querier needs to be in `ContextData`.

But there is a simpler way to temporarily access the querier without all this move in, move out action:

```rust
pub fn with_querier_from_context<S: Storage, Q: Querier, F: FnMut(&Q)>(ctx: &Ctx, mut func: F) {
    let context_data = unsafe { Box::from_raw(ctx.data as *mut ContextData<S, Q>) };
    if let Some(q) = &context_data.querier {
        func(q);
    }
    Box::leak(context_data);
}
```

Reply:
The problem with this last one is a bug you found a while ago.
What if this callback calls `assert!()` and fails (which is the same as `panic!()`)

In this case, it unwinds the stack and first drops the local variables, (incl `q` and `context_data`), then goes further up the stack and calls the destructor (`destroy unmanaged storage`). Which will free out context_data again.

My approach was to then leave context_data full of `None` while we had the data in local variables (so they only got cleared once on panic) and then reset them when finishing. I think you made a test for this, but we should make more.

Reply:
> The only place for long living data is Ctx.data, so the querier needs to be in ContextData.

Exactly. If there is no data (or it is copyable), there is no problem.